### PR TITLE
qt6

### PR DESCRIPTION
**Pull request by looooo**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
@conda-forge/core do you know what this means?

```
To use a cross-compiled Qt, please set the QT_HOST_PATH cache variable to
  the location of your host Qt installation.
```

next question (osx):

```
ld: framework not found Metal
```

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
does this help
https://github.com/conda-forge/qt-wayland-feedstock/blob/qt6/recipe/build.sh#L14

File: recipe/build.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ cmake -G "Ninja" \
       -D CMAKE_INSTALL_PREFIX=$PREFIX \
       -D CMAKE_PREFIX_PATH=$PREFIX \
       -D CMAKE_BUILD_TYPE="Release" \
-      -D SOQT_USE_QT5=ON \
+      -D SOQT_USE_QT6=ON \
```
Comment (hmaarrfk) [inline recipe/build.sh]:
```suggestion
      -D SOQT_USE_QT6=ON \
      -D QT_HOST_PATH=${PREFIX} \
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -26,14 +30,15 @@ requirements:
         - {{ cdt('libxdamage') }}  # [linux]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:29]: these x packages should be made conda-forge packages.

**Comment by hmaarrfk** [comment]
fyi, we have been working to make qt6 and qt5 libraries co-installable to ease with the transition.

are the library names the same? or do the qt6 ones include a "6" inside the name?

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
-{% set sha256 = "9f535af59f07c907022815679681bd345c9dec2f408c74833f6b1b24bca40e1f" %}
 
 package:
     name: {{ name }}
     version: {{ version }}
 
 source:
-    - url: https://github.com/coin3d/soqt/releases/download/SoQt-{{ version }}/soqt-{{ version }}-src.tar.gz
-      sha256: {{ sha256 }}
+    - url: https://github.com/coin3d/soqt/archive/ab0595fe864470451ca303afb9b49c796daf2c00.tar.gz
```
Comment (hmaarrfk) [inline recipe/meta.yaml]: this is very strange, you are building an unversioned version of soqt. why is this?

Comment (looooo) [reply recipe/meta.yaml]: Yes qt6 support is not in the last available version

Comment (hmaarrfk) [reply recipe/meta.yaml]: is your goal to publish this unreleased version in conda-forge?

Comment (looooo) [reply recipe/meta.yaml]: I will ask if it is possible to publish a new version.

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
Just a small note:

When i meant co-installable, i really meant:
- same environment.
- install both packages
- no symbol clash
- no library clash

it seems that this library will always use the filename `$PREFIX/lib/libSoQt.so.20.6.0` for both qt6 and qt5.

**Comment by hmaarrfk** [comment]
so maybe making it co-installable is too hard for this particular package.

**Comment by looooo** [comment]
https://github.com/coin3d/soqt/issues/74

**Comment by looooo** [comment]
Ref https://github.com/coin3d/soqt/issues/74

In my mind it's not necessary to have qt6 in the same env as qt5 so I am not going to support this idea of having soqt-qt6 and soqt-qt5 in the same env.

There is still one issue with windows. I need to do something similar as I did in the build.sh script
https://github.com/looooo/soqt-feedstock/blob/qt6/recipe/build.sh#L19

But not yet sure how translate this to the batch file

Ok this seems to work. Maybe you can have a look if it makes sense:
https://github.com/looooo/soqt-feedstock/blob/qt6/recipe/bld.bat#L5

**Comment by hmaarrfk** [comment]
Just a note of what I found

i see that the last release was 2020, and that the CMake changes where made in 2021.
https://github.com/coin3d/soqt/commit/b519a47f45b8b7cc7027e3f139dd9e6a922e1c2f

**Comment by looooo** [comment]
@hmaarrfk is there still the option to avoid cross-compiling with azure? Iirc there was such an option back than by using this:
```
build_platform:
  linux_aarch64: linux_aarch64
  osx_arm64: osx_arm64
```

With this osx-arm64 is skipped. So I guess there is no way to not use cross-compilation, right?

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by hmaarrfk** [comment]
you must cross compile for osx.

I think cross compiling should work

qt6-charts is cross compiled:
https://github.com/conda-forge/qt-charts-feedstock/blob/main/recipe/build.sh

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
> you must cross compile for osx.
> 
> I think cross compiling should work
> 
> qt6-charts is cross compiled: https://github.com/conda-forge/qt-charts-feedstock/blob/main/recipe/build.sh

I still see this issue with qt6:

```
$PREFIX/lib/qt6/moc: Bad CPU type in executable
```

Maybe the the host qt is used here?

**Comment by hmaarrfk** [comment]
yes it seems like it is. it hsould be using it from ${BUILD_PREFIX}

**Comment by hmaarrfk** [comment]
It might be dumb, but try something like:

```
rm -f ${PREFIX}/lib/qt6/moc
ln -s ${BUILD_PREFIX}/lib/qt6/moc ${PREFIX}/lib/qt6/moc
```
for osx.

**Comment by hmaarrfk** [comment]
if that works, we should report to `qt6-main` feedstock.

File: recipe/build.sh [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
-declare -a QT_SELECTOR
+QT_SELECTOR=()
 
 if [[ "$build_variant" = "qt5" ]]; then
-      QT_SELECTOR+=(-D SOQT_USE_QT5=ON)
+      QT_SELECTOR+=("-D SOQT_USE_QT5=ON")
 fi
 if [[ "$build_variant" = "qt6" ]]; then
-      QT_SELECTOR+=(-D SOQT_USE_QT6=ON)
+      QT_SELECTOR+=("-D SOQT_USE_QT6=ON")
```
Comment (hmaarrfk) [inline recipe/build.sh]: honestly, why bother with bash arrays...

```
QT_SELECTION=-DSOQT_USE_QT6=ON
```

then
```
cmake \
   ${QT_SELECTOR} \
```

Comment (looooo) [reply recipe/build.sh]: thanks, yes this is better

**Comment by looooo** [comment]
> It might be dumb, but try something like:
> 
> ```
> rm -f ${PREFIX}/lib/qt6/moc
> ln -s ${BUILD_PREFIX}/lib/qt6/moc ${PREFIX}/lib/qt6/moc
> ```
> 
> for osx.

the qt-main dependency was missing because of an error in the meta.yaml file. Hope this fixes the issue now.

**Comment by hmaarrfk** [comment]
My suggestion to remove the binary didn't come out of nowhere, I have both qt5-main and qt6-main and I can somewhat see that there are differences between the MOC definition:

```
/lib/cmake 
14:48 $ rg "::moc"
Qt6Core/Qt6CoreMacros.cmake
127:                       COMMAND ${QT_CMAKE_EXPORT_NAMESPACE}::moc ${_moc_extra_parameters_file}
1150:        DEPENDS ${QT_CMAKE_EXPORT_NAMESPACE}::moc ${automoc_dependencies} ${manual_dependencies}
1151:        COMMAND ${QT_CMAKE_EXPORT_NAMESPACE}::moc

Qt5Core/Qt5CoreConfigExtras.cmake
16:if (NOT TARGET Qt5::moc)
17:    add_executable(Qt5::moc IMPORTED)
22:    set_target_properties(Qt5::moc PROPERTIES
26:    get_target_property(QT_MOC_EXECUTABLE Qt5::moc LOCATION)
43:set(Qt5Core_MOC_EXECUTABLE Qt5::moc)

Qt6CoreTools/Qt6CoreToolsConfig.cmake
60:    __qt_internal_promote_target_to_global(Qt6::moc)
98:set(Qt6CoreTools_TARGETS "Qt6::syncqt;Qt6::moc;Qt6::rcc;Qt6::tracepointgen;Qt6::tracegen;Qt6::cmake_automoc_parser;Qt6::qlalr;Qt6::qtpaths;Qt6::androiddeployqt;Qt6::androidtestrunner;Qt6::qmake")

Qt6CoreTools/Qt6CoreToolsAdditionalTargetInfo.cmake
26:get_target_property(_qt_imported_location Qt6::moc IMPORTED_LOCATION_RELEASE)
27:get_target_property(_qt_imported_location_default Qt6::moc IMPORTED_LOCATION_${QT_DEFAULT_IMPORT_CONFIGURATION})
29:# Import target "Qt6::moc" for configuration "RelWithDebInfo"
30:set_property(TARGET Qt6::moc APPEND PROPERTY IMPORTED_CONFIGURATIONS RELWITHDEBINFO)
33:    set_property(TARGET Qt6::moc PROPERTY IMPORTED_LOCATION_RELWITHDEBINFO "${_qt_imported_location}")
36:# Import target "Qt6::moc" for configuration "MinSizeRel"
37:set_property(TARGET Qt6::moc APPEND PROPERTY IMPORTED_CONFIGURATIONS MINSIZEREL)
40:    set_property(TARGET Qt6::moc PROPERTY IMPORTED_LOCATION_MINSIZEREL "${_qt_imported_location}")
45:    set_property(TARGET Qt6::moc PROPERTY IMPORTED_LOCATION "${_qt_imported_location_default}")

Qt6CoreTools/Qt6CoreToolsTargets.cmake
22:foreach(_cmake_expected_target IN ITEMS Qt6::syncqt Qt6::moc Qt6::rcc Qt6::tracepointgen Qt6::tracegen Qt6::cmake_automoc_parser Qt6::qlalr Qt6::qtpaths Qt6::androiddeployqt Qt6::androidtestrunner Qt6::qmake)
68:# Create imported target Qt6::moc
69:add_executable(Qt6::moc IMPORTED)
71:set_target_properties(Qt6::moc PROPERTIES

Qt6CoreTools/Qt6CoreToolsTargets-release.cmake
17:# Import target "Qt6::moc" for configuration "Release"
18:set_property(TARGET Qt6::moc APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
19:set_target_properties(Qt6::moc PROPERTIES
23:list(APPEND _cmake_import_check_targets Qt6::moc )
24:list(APPEND _cmake_import_check_files_for_Qt6::moc "${_IMPORT_PREFIX}/lib/qt6/moc" )

Qt6/QtToolHelpers.cmake
592:        # Otherwise targets like Qt6::moc are not available in a top-level cross-build.

Qt6/QtAutogenHelpers.cmake
152:                           COMMAND ${QT_CMAKE_EXPORT_NAMESPACE}::moc "@${moc_parameters_file}"
153:                           DEPENDS "${infile}" ${moc_depends} ${QT_CMAKE_EXPORT_NAMESPACE}::moc
```

**Comment by hmaarrfk** [comment]
one problem with this new recipe, is that you will effectively force everybody that depends on `soqt` to update to qt6. I just don't think the world is ready for this yet.

```
# soqt6 and soqt (for qt5) are mutually exclusivie
name: soqt6   # [build_variant == "qt6"]
name: soqt    # [build_variant == "qt5"]

# [...]

requirements:
  run_constrained:
    - soqt <0     # [build_variant == "qt6"]
    - soqt6 <0    # [build_variant == "qt5"]
```

This is just an idea, but i'm not too sure how to transition packages like this. I have the same problem with opencv....

**Comment by looooo** [comment]
> one problem with this new recipe, is that you will effectively force everybody that depends on `soqt` to update to qt6. I just don't think the world is ready for this yet.
> 
> ```
> # soqt6 and soqt (for qt5) are mutually exclusivie
> name: soqt6   # [build_variant == "qt6"]
> name: soqt    # [build_variant == "qt5"]
> 
> # [...]
> 
> requirements:
>   run_constrained:
>     - soqt <0     # [build_variant == "qt6"]
>     - soqt6 <0    # [build_variant == "qt5"]
> ```
> 
> This is just an idea, but i'm not too sure how to transition packages like this. I have the same problem with opencv....

We are building here for qt5 and qt6. If an environment needs qt5, conda should not install soqt depending on qt6. This is my understanding. I am a fan of "only one version of a library in one environment". Isn't this the main feature of virtual environments to not have to deal with these issues coming from having multiple versions of the same library in one environment. 

Not sure I understand why qt6 needs to have its own name. This is also not done for python for example, why should we do this with qt?

**Comment by hmaarrfk** [comment]
> We are building here for qt5 and qt6. If an environment needs qt5, conda should not install soqt depending on qt6. This is my understanding. I am a fan of "only one version of a library in one environment". Isn't this the main feature of virtual environments to not have to deal with these issues coming from having multiple versions of the same library in one environment.

are you able to compile this package yourself for the machine you are currently working on?

If so, I suggest that you attempt to build both the packages locally, and upload them to your personal anaconda channel.

Then add your channel.

```
# ok my channel has the linux packages, this will mess with your stuff since i have a few very experimental packages
# i built them just now
# mamba config --add channels mark.harfouche  
# ok don't ask me why adding the channel twice makes a difference (i don't really know), lets just roll with it.
CHANNEL_ARGS="--channel mark.harfouche --channel conda-forge --channel mark.harfouche --override-channels"
mamba create --name soqt_test qt-main python=3.10 pyqt  ${CHANNEL_ARGS}
mamba activate soqt_test
mamba install soqt ${CHANNEL_ARGS}
```


```
$ mamba install soqt ${CHANNEL_ARGS}

Looking for: ['soqt']

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache
mark.harfouche/noarch                                         No change
mark.harfouche/linux-64                                       No change

Pinned packages:
  - python 3.10.*


Transaction

  Prefix: /h/mark/miniforge3/envs/soqt_test

  Updating specs:

   - soqt
   - ca-certificates
   - openssl


  Package              Version  Build          Channel              Size
──────────────────────────────────────────────────────────────────────────
  Install:
──────────────────────────────────────────────────────────────────────────

  + double-conversion    3.3.0  h59595ed_0     conda-forge        Cached
  + libwebp-base         1.3.2  hd590300_0     conda-forge        Cached
  + libpciaccess          0.17  h166bdaf_0     conda-forge        Cached
  + lerc                 4.0.0  h27087fc_0     conda-forge        Cached
  + libdeflate            1.19  hd590300_0     conda-forge        Cached
  + xcb-util-cursor      0.1.4  hd590300_1     conda-forge        Cached
  + xorg-libxt           1.3.0  hd590300_1     conda-forge        Cached
  + libglu               9.0.0  hac7e632_1003  conda-forge        Cached
  + xorg-inputproto      2.3.2  h7f98852_1002  conda-forge        Cached
  + xorg-fixesproto        5.0  h7f98852_1002  conda-forge        Cached
  + libdrm             2.4.114  h166bdaf_0     conda-forge        Cached
  + libtiff              4.6.0  ha9c0a0a_2     conda-forge        Cached
  + xorg-libxfixes       5.0.3  h7f98852_1004  conda-forge        Cached
  + qt6-main             6.5.3  h450cf1c_0     conda-forge        Cached
  + xorg-libxi          1.7.10  h7f98852_0     conda-forge        Cached
  + coin3d               4.0.0  ha2d195d_1013  conda-forge        Cached
  + soqt                 1.6.0  h9bf6919_4     mark.harfouche     Cached

  Summary:

  Install: 17 packages

  Total download: 0 B

──────────────────────────────────────────────────────────────────────────



Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done

```

```
# packages in environment at /h/mark/miniforge3/envs/soqt_test:
#
# Name                    Version                   Build  Channel
pyqt                      5.15.9          py310h04931ad_5    conda-forge
pyqt5-sip                 12.12.2         py310hc6cd4ac_5    conda-forge
qt-main                   5.15.8              h82b777d_17    conda-forge
qt6-main                  6.5.3                h450cf1c_0    conda-forge
soqt                      1.6.0                h9bf6919_4    mark.harfouche
```

[soqt_mamba_list.txt](https://github.com/conda-forge/soqt-feedstock/files/13329061/soqt_mamba_list.txt)


so in other words, conda isn't smart enough to "prioritize the current version of qt" because it doesnt' know that "qt5-main" and "qt6-main" are similar.


While your ambitions of "this is why virtual environments exist" is pretty correct, the problem is that qt is notoriously complex. The qt4->qt5 transition took years, and the qt5->qt6 is still ongoing for many packages.


Much of the discussion can be found:
https://github.com/conda-forge/qt-main-feedstock/issues/49

I guess for now, I would suggest to follow the trend: adding a "6" to your package name.

The exclusion helps ensure that you don't install the `soqt6` and the `soqt` (5) package together. but it will ensure that when you want to install the qt6 variant, you get what you expect.

**Comment by looooo** [comment]
thanks for clarification. I will try the soqt6 approach.
actually I need soqt with qt6 for pivy. Because I want to add python3.12 support. This is needed because pyside2 is not available for py312 but pyside6 is. Not sure if there will be a pyside2 python312 build available soon.

**Comment by hmaarrfk** [comment]
you have an other issue where pivy has dependencies on PySide2.

I suggestion moving to qtpy, https://github.com/conda-forge/qtpy-feedstock

then you don't need to worry too too much about pyside2 vs pyside6

for the most part

```
sed "s/PySide2/qtpy" -i filename.py
```
does the trick.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,21 +1,25 @@
 {% set name = "soqt" %}
 {% set version = "1.6.0" %}
-{% set sha256 = "9f535af59f07c907022815679681bd345c9dec2f408c74833f6b1b24bca40e1f" %}
 
 package:
```
Comment (hmaarrfk) [inline recipe/meta.yaml:10]:
```suggestion
{% if build_variant == "qt6" %}
{% set name_suffix = "6" %}
{% else %}
{% set name_suffix = "" %}
{% endif %}

package:
```

then 
```
name: soqt{{ name_suffix }}
```

sigh, seems like we need to trick the template engine a little bit.

Comment (hmaarrfk) [reply recipe/meta.yaml:10]: not sure if the above helps, but I would try that.

**Comment by looooo** [comment]
strange your suggestion with naming the packages "soqt6" and "soqt" did not work. I get this:
```
Error: package/name missing in: '/home/conda/recipe_root/meta.yaml'
```

**Comment by hmaarrfk** [comment]
I still don't know a good resolution for the fact that you don't have an official version.

It seems that the number of commits isn't that high.

Maybe one possibility is to cherry pick what you need?

Or just use patches for qt6 variant?

If the patches really only affect the build files, it is easier to audit, and to ensure that certain things don't materially change with the library.

**Comment by hmaarrfk** [comment]
> strange your suggestion with naming the packages "soqt6" and "soqt" did not work. I get this:

my first suggestion was to have a ABI migration branch ;)

**Comment by looooo** [comment]
> I still don't know a good resolution for the fact that you don't have an official version.
> 
> It seems that the number of commits isn't that high.
> 
> Maybe one possibility is to cherry pick what you need?
> 
> Or just use patches for qt6 variant?
> 
> If the patches really only affect the build files, it is easier to audit, and to ensure that certain things don't materially change with the library.

actually I don't think it's that bad. Soqt is only used with pivy, and I don't think anyone is using it actually as there is a newer replacement available (quarter). Also pivy is only used with freecad. And freecad is not using soqt...

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,21 +1,30 @@
...
+
+{% if build_variant == "qt6" %}
+{% set name_suffix = "6" %}
+{% else %}
+{% set name_suffix = "" %}
+{% endif %}
 
 package:
-    name: {{ name }}
+    name: {{ name }} {{ name_suffix }}
```
Comment (hmaarrfk) [inline recipe/meta.yaml]:
```suggestion
    name: {{ name }}{{ name_suffix }}
```

**Comment by hmaarrfk** [comment]
> actually I don't think it's that bad. Soqt  is only used with pivy,

I don't really like this line of thinking. 

The goal of conda-forge is to build out a cohesive ecosystem of packages that are suitable for a wider audience.

I don't think that releasing experimental versions is ok anymore....
I shared my thoughts in the past in https://github.com/conda-forge/sqlacodegen-feedstock/issues/11#issuecomment-1433312739

It is unclear to me why instead of going from a single tar gz you now have to download 3 difference sources. This seems to indicate a major change in the library.

Maybe for python 3.12 support, you should focus your efforts on moving to `quarter`.

**Comment by looooo** [comment]
> > actually I don't think it's that bad. Soqt  is only used with pivy,
> 
> I don't really like this line of thinking.
> 
> The goal of conda-forge is to build out a cohesive ecosystem of packages that are suitable for a wider audience.
> 
> I don't think that releasing experimental versions is ok anymore.... I shared my thoughts in the past in [conda-forge/sqlacodegen-feedstock#11 (comment)](https://github.com/conda-forge/sqlacodegen-feedstock/issues/11#issuecomment-1433312739)
> 
> It is unclear to me why instead of going from a single tar gz you now have to download 3 difference sources. This seems to indicate a major change in the library.
> 
> Maybe for python 3.12 support, you should focus your efforts on moving to `quarter`.

yes I agree with you, that this is not good practice. The best solution is a new version of the library. But with these old libraries it's difficult to get this. See for example here: 
https://github.com/coin3d/coin/issues/489
this is already one year old.

**Comment by hmaarrfk** [comment]
I'm not sure why you are being very selective about what you are reading from my messages.

I didn't say this was outright not ok, but I am suggesting that you pull specific patches that you need.

We did this for vigra for a long time until I managed to get involved. This was our great cleanup.
https://github.com/conda-forge/vigra-feedstock/pull/121

This just really helps for transparency to other potential users.

This is just the tradeoff of working in larger ecosystem.

**Comment by hmaarrfk** [comment]
new version!!!!!

**Pull request merged**